packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts

```diff
@@ -129,6 +129,21 @@ export default createRule({
           caseTypes.add(getNodeType(switchCase.test));
         }
 
+        const isFiniteType = unionTypes.every(isLiteralType);
+        const isSymbolType = unionTypes.some(type =>
+          isTypeFlagSet(type, ts.TypeFlags.UniqueESSymbol),
+        );
+        const isStringLiteralOrBoolean = unionTypes.some(type =>
+          isTypeFlagSet(
+            type,
+            ts.TypeFlags.BooleanLiteral || ts.TypeFlags.StringLiteral,
+          ),
+        );
+
+        if (!isFiniteType && !isSymbolType && !isStringLiteralOrBoolean) {
```

Comment:
[Performance] This probably will never be a real issue, but - you can short circuit each of these cases to not have to go through later checks after an earlier one is `true`.

```ts
if (unionTypes.every((isLiteralType)) {
  return;
}

if (unionTypes.some( ... )) {
  return;
}
```

We've been bit in the past by doing unnecessary work that _was_ fine when written but ended up being a lot of extra cycles.